File: app/src/main/java/us/kesslern/freshms/MainActivity.java
package us.kesslern.freshms;

import android.Manifest;
import android.content.ComponentName;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.v4.app.ActivityCompat;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.CompoundButton;
import android.widget.Switch;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = MainActivity.class.getSimpleName();
    private int sent = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        ActivityCompat.requestPermissions(this,
                new String[]{Manifest.permission.RECEIVE_SMS},
                12);

        Switch mySwitch = (Switch) findViewById(R.id.switchSync);
        mySwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                if (isChecked) {
                    enableBroadcastReceiver();
                } else {
                    disableBroadcastReceiver();
                }
            }
        });
    }

    public void update(View view) {
        sent++;
        setSentText();
    }

    public void enableBroadcastReceiver() {
        ComponentName receiver = new ComponentName(this, SMSReceiver.class);
        PackageManager pm = this.getPackageManager();

        pm.setComponentEnabledSetting(receiver,
                PackageManager.COMPONENT_ENABLED_STATE_ENABLED,
                PackageManager.DONT_KILL_APP);
        Log.v(TAG, "Enabled broadcast receiver");
    }

    public void disableBroadcastReceiver() {
        ComponentName receiver = new ComponentName(this, SMSReceiver.class);
        PackageManager pm = this.getPackageManager();

        pm.setComponentEnabledSetting(receiver,
                PackageManager.COMPONENT_ENABLED_STATE_DISABLED,
                PackageManager.DONT_KILL_APP);
        Log.v(TAG, "Disabled broadcst receiver");
    }

    private void setSentText() {
        TextView textView = (TextView) findViewById(R.id.totalSent);
        textView.setText("Total sent: " + sent);
    }
}

File: app/src/main/java/us/kesslern/ascient/MainActivity.kt
package us.kesslern.ascient

import android.content.Context
import android.content.IntentFilter
import android.content.SharedPreferences
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.Log
import com.github.kittinunf.fuel.core.FuelManager
import kotlinx.android.synthetic.main.activity_main.*
import java.util.*

class MainActivity : AppCompatActivity() {

    private var totalReceived = 0
    private val preferences: SharedPreferences by lazy { getPreferences(Context.MODE_PRIVATE) }
    private val intentFilter = IntentFilter("android.provider.Telephony.SMS_RECEIVED")
    private val smsReceiver = SMSBroadcastReceiver(this)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        PermissionHandlerService.requestPermissions(this)
        FuelManager.instance.baseHeaders = mapOf(
                "Content-Type" to "application/json",
                "Accept" to "application/json")

        if (!preferences.contains("applicationId")) {
            preferences.edit().putString("applicationId", UUID.randomUUID().toString()).apply()
        }
        val applicationId = preferences.getString("applicationId", "")

        totalSent.setText(R.string.no_messages_received)
        uuid.text = applicationId

        loginButton.setOnClickListener {
            Log.i(TAG, "${usernameInput.text}, ${passwordInput.text}")
            ApiService.login(usernameInput.text.toString(), passwordInput.text.toString(),
                    { Log.i(TAG, "Success! $it")},
                    { Log.i(TAG, "Failure!")})
        }

        switchSync.setOnCheckedChangeListener { _, isChecked ->
            updateBroadcastReceiver(isChecked)
        }
        switchSync.isChecked = preferences.getBoolean(ENABLED_SWITCH_STATE, false)
    }

    fun update() {
        totalReceived++
        totalSent.text = String.format(
                getString(R.string.total_received),
                totalReceived.toString())
    }

    private fun updateBroadcastReceiver(enabled: Boolean) {
        if (enabled) {
            registerReceiver(smsReceiver, intentFilter)
        } else {
            unregisterReceiver(smsReceiver)
        }

        preferences.edit().putBoolean(ENABLED_SWITCH_STATE, enabled).apply()
        Log.d(TAG, tag(SMSBroadcastReceiver::class) + " state :" + enabled.toString())
    }

    companion object {
        private val TAG = tag(MainActivity::class)
        private val ENABLED_SWITCH_STATE = "switch_state"
    }
}
